pkg/tfbridge/validate_input_types.go

```diff
@@ -35,215 +35,249 @@ func NewInputValidator(urn resource.URN, schema pschema.PackageSpec) *PulumiInpu
+		(propertyValue.IsOutput() && !propertyValue.OutputValue().Known) {
 		return nil
 	}
 
-	// handle non-required properties that are null
-	if propertyValue.IsNull() {
-		if slices.Contains(requiredProps, propertyKey) {
-			return []TypeFailure{
-				{
-					ResourcePath: propertyPath.String(),
-					Reason:       fmt.Sprintf("property %s is required", propertyKey),
-				},
-			}
-		}
+	// for known first-class outputs simply validate their known value
+	if propertyValue.IsOutput() && propertyValue.OutputValue().Known {
+		elementValue := propertyValue.OutputValue().Element
+		return v.validatePropertyValue(elementValue, typeSpec, propertyPath)
 	}
 
-	failures := []TypeFailure{}
-	if propertyKey != "" {
-		propertyPath = append(propertyPath, propertyKey)
+	// for secrets validate their inner value
+	if propertyValue.IsSecret() {
+		elementValue := propertyValue.SecretValue().Element
+		return v.validatePropertyValue(elementValue, typeSpec, propertyPath)
 	}
 
-	// perform some initial easy type matching. fail fast.
-	possible, ok := v.matchType(propertyValue, typeSpec)
-	if !ok {
-		return []TypeFailure{
-			{
+	// now we are going to switch on the desired type
+	//
+	// a good reference here for the semantics of the TypeSpec is
+	// https://github.com/pulumi/pulumi/blob/master/pkg/codegen/schema/bind.go#L881
+	//
+	// this code is not using the binder functionality as that would require a plugin loader to resolve references,
+	// which is not desirable for the provider runtime
+	//
+	// possibly this could be done in the future with a way to still use the bind code while resolving references to
+	// unchecked types as using the bound representation would make the code cleaner
+	//
+	// for now the code simply follows bindSpec
+
+	if typeSpec.Ref != "" {
+		objType := v.getType(typeSpec.Ref)
+		// Refusing to validate unknown or unresolved type.
+		if objType == nil {
+			return nil
+		}
+
+		if !propertyValue.IsObject() {
+			return []TypeFailure{{
 				ResourcePath: propertyPath.String(),
 				Reason: fmt.Sprintf(
-					"expected %s type, got %s type",
-					possible.toString(),
+					"expected an object, got %s type",
 					propertyValue.TypeString(),
 				),
-			},
+			}}
 		}
+
+		return v.validatePropertyMap(
+			propertyValue.ObjectValue(),
+			objType.Required,
+			objType.Properties,
+			propertyPath,
+		)
 	}
 
-	// array type
-	if propertyValue.IsArray() {
-		for idx, arrayValue := range propertyValue.ArrayValue() {
-			pb := append(propertyPath, idx)
-			if typeSpec.Items != nil {
-				failure := v.validateTypeSpec("", arrayValue, *typeSpec.Items, pb, requiredProps)
-				if failure != nil {
-					failures = append(failures, failure...)
-				}
-			} else if len(typeSpec.OneOf) > 0 {
-				failure := v.validateOneOf("", arrayValue, typeSpec.OneOf, pb)
-				if failure != nil {
-					failures = append(failures, failure...)
-				}
-			}
-		}
+	if typeSpec.OneOf != nil {
+		// TODO type check against union types.
```

Comment:
This is important case for maps and need to be filled in, should be trivial just omitting for lack of time.